mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -0,0 +1,560 @@
+package scala.meta.internal.pc
+import java.util
+import java.util.logging.Logger
+import java.{util => ju}
+
+import scala.collection.mutable.ListBuffer
+import scala.reflect.internal.util.Position
+
+import scala.meta.internal.jdk.CollectionConverters._
+import scala.meta.pc.VirtualFileParams
+import scala.meta.tokens._
+
+import org.checkerframework.common.returnsreceiver.qual.This
+import org.eclipse.lsp4j.SemanticTokenModifiers
+import org.eclipse.lsp4j.SemanticTokenTypes
+
+/**
+ * Corresponds to tests.SemanticHighlightLspSuite
+ */
+final class SemanticTokenProvider(
+    protected val cp: MetalsGlobal // compiler
+    ,
+    val params: VirtualFileParams,
+    val capableTypes: util.List[String],
+    val capableModifiers: util.List[String]
+) {
+
+  // alias for long notation
+  def getTypeId(p: String): Int = capableTypes.indexOf(p)
+
+  // log tools
+  val logger: Logger = Logger.getLogger(classOf[This].getName)
+  val strSep = ", "
+  val linSep = "\n"
+
+  // initialize semantic tree
+  val unit: cp.RichCompilationUnit = cp.addCompilationUnit(
+    params.text(),
+    params.uri().toString(),
+    None
+  )
+  cp.typeCheck(unit) // initializing unit
+  val (root, source) = (unit.lastBody, unit.source)
+
+  def unitPos(offset: Int): Position = unit.position(offset)
+  val nodes: Set[NodeInfo] = traverser.traverse(Set.empty[NodeInfo], root)
```

Comment:
```suggestion
  val nodes: Set[NodeInfo] = traverser.traverse(Set.empty[NodeInfo], root).sortBy(_.position.start)
```
we could then just iterate over the nodes, so that would be more effective and we would go through the list only once instead of every time we want to get info about a node.